### ISSUE TITLE
Max local stream count exceeded for HttpClient with HTTP/2 transport  

### ISSUE THREAD

**Issue by sbordet** [eclipse/jetty.project#2672]
```
java.lang.IllegalStateException: Max local stream count 128 exceeded
	at org.eclipse.jetty.http2.HTTP2Session.createLocalStream(HTTP2Session.java:696)
	at org.eclipse.jetty.http2.HTTP2Session.newStream(HTTP2Session.java:491)
	at org.eclipse.jetty.http2.client.http.HttpSenderOverHTTP2.sendHeaders(HttpSenderOverHTTP2.java:93)
	at org.eclipse.jetty.client.HttpSender.send(HttpSender.java:212)
	at org.eclipse.jetty.http2.client.http.HttpChannelOverHTTP2.send(HttpChannelOverHTTP2.java:95)
	at org.eclipse.jetty.client.HttpChannel.send(HttpChannel.java:128)
	at org.eclipse.jetty.client.HttpConnection.send(HttpConnection.java:201)
	at org.eclipse.jetty.http2.client.http.HttpConnectionOverHTTP2.send(HttpConnectionOverHTTP2.java:72)
	at org.eclipse.jetty.http2.client.http.HttpDestinationOverHTTP2.send(HttpDestinationOverHTTP2.java:38)
	at org.eclipse.jetty.client.HttpDestination.process(HttpDestination.java:347)
	at org.eclipse.jetty.client.HttpDestination.process(HttpDestination.java:305)
	at org.eclipse.jetty.client.HttpDestination.send(HttpDestination.java:295)
	at org.eclipse.jetty.client.HttpDestination.send(HttpDestination.java:270)
	at org.eclipse.jetty.client.HttpDestination.send(HttpDestination.java:247)
	at org.eclipse.jetty.client.HttpClient.send(HttpClient.java:573)
	at org.eclipse.jetty.client.HttpRequest.send(HttpRequest.java:727)
	at org.eclipse.jetty.client.HttpRequest.send(HttpRequest.java:719)
```

This exception should not happen while using `HttpClient`: requests should be queued instead, but it has been seen under high load, greater than 6 K req/s.

There must be some misalignment between the `ConnectionPool`, which is obviously giving out a connection thinking there is room for a request, and `HTTP2Session` thinking there are too many streams for that connection/session.

**Comment by sbordet** [comment]
Nope, this issue is for HTTP/2, `ConnectionPoolTest` is only for HTTP/1.1.

**Comment by joakime** [comment]
Is this related to issue #2349 ?

**Comment by sbordet** [comment]
No, it's not related.

**Comment by sbordet** [comment]
This issue is due to the fact that there is a race between the establishment of a connection, a client sending many requests concurrently, and the server sending to the client the `max_concurrent_streams` setting.

When the connection is established, the client preface is sent; when the write of the client preface succeeds, the connection is made available to the `ConnectionPool`.
A client that is sending many concurrent requests can now use this connection to send N requests, where N is the default multiplex value, which is by default unlimited.

Meanwhile, the server receives the client preface and sends to the client its preface, that contains `max_concurrent_streams=M` where `M<N`.

This condition may already trigger an exception on the server, as the client may have in-flight a number of concurrent requests that is already greater than `M`.

When the client receives the server preface, it first update the `HTTP2Session` and only after updates the `ConnectionPool`. 
This leaves a window of time open for the client to send requests that, according to the `ConnectionPool` that has not been updated yet with the smaller value, can be multiplexed in a single connection; however, when the streams correspondent to the requests are created by the `HTTP2Session` they will exceed `max_concurrent_streams` because the `HTTP2Session` has already been updated with the smaller value.

**Comment by gregw** [comment]
@sbordet waiting for the settings frame will add unnecessary latency to the connection process.   So I'm wondering if we should instead allow an initial assumed window size to be configured.  If it is 0, then you get the behaviour you have implemented in 9aa18c4, if it is >0, then that number of streams can be created prior to the settings frame arriving.... hmmm or will that just cause multiple connections to be created instead???
